actors/market/tests/market_actor_test.rs

```diff
@@ -1278,6 +1279,295 @@ fn active_deals_multiple_times_with_different_providers() {
+        result: sig_result,
+    });
+
+    let params: PublishStorageDealsParams = PublishStorageDealsParams {
+        deals: vec![ClientDealProposal { proposal: deal_proposal, client_signature: sig }],
+    };
+
+    assert_eq!(
+        exit_code,
+        rt.call::<MarketActor>(
+            Method::PublishStorageDeals as u64,
+            &RawBytes::serialize(params).unwrap(),
+        )
+        .unwrap_err()
+        .exit_code()
+    );
+    rt.verify();
+    // TODO: actor.checkState(rt)
+}
+
+#[cfg(test)]
+mod publish_storage_deals_failures {
+    use super::*;
+
+    #[test]
+    fn deal_end_after_deal_start() {
+        let f = |_rt: &mut MockRuntime, d: &mut DealProposal| {
+            d.start_epoch = 10;
+            d.end_epoch = 9;
+        };
+        assert_deal_failure(true, f, ExitCode::USR_ILLEGAL_ARGUMENT, Ok(()));
+    }
+
+    #[test]
+    fn current_epoch_greater_than_start_epoch() {
+        let current_epoch = ChainEpoch::from(5);
+
+        let f = |_rt: &mut MockRuntime, d: &mut DealProposal| {
+            d.start_epoch = current_epoch - 1;
+        };
+        assert_deal_failure(true, f, ExitCode::USR_ILLEGAL_ARGUMENT, Ok(()));
+    }
+
+    #[test]
+    fn deal_duration_greater_than_max_deal_duration() {
+        let f = |_rt: &mut MockRuntime, d: &mut DealProposal| {
+            d.start_epoch = ChainEpoch::from(10);
+            d.end_epoch = d.start_epoch + (540 * EPOCHS_IN_DAY) + 1
+        };
+        assert_deal_failure(true, f, ExitCode::USR_ILLEGAL_ARGUMENT, Ok(()));
+    }
+
+    #[test]
+    fn negative_price_per_epoch() {
+        let f = |_rt: &mut MockRuntime, d: &mut DealProposal| {
+            d.storage_price_per_epoch = TokenAmount::from(-1);
+        };
+        assert_deal_failure(true, f, ExitCode::USR_ILLEGAL_ARGUMENT, Ok(()));
+    }
+
+    #[test]
+    fn price_per_epoch_greater_than_total_filecoin() {
+        let f = |_rt: &mut MockRuntime, d: &mut DealProposal| {
+            d.storage_price_per_epoch = TOTAL_FILECOIN.clone() + 1;
+        };
+        assert_deal_failure(true, f, ExitCode::USR_ILLEGAL_ARGUMENT, Ok(()));
+    }
+
+    #[test]
+    fn negative_provider_collateral() {
+        let f = |_rt: &mut MockRuntime, d: &mut DealProposal| {
+            d.provider_collateral = TokenAmount::from(-1);
+        };
+        assert_deal_failure(true, f, ExitCode::USR_ILLEGAL_ARGUMENT, Ok(()));
+    }
+
+    #[test]
+    fn provider_collateral_greater_than_max_collateral() {
+        let f = |_rt: &mut MockRuntime, d: &mut DealProposal| {
+            d.provider_collateral = TOTAL_FILECOIN.clone() + 1;
+        };
+        assert_deal_failure(true, f, ExitCode::USR_ILLEGAL_ARGUMENT, Ok(()));
+    }
+
+    #[test]
+    fn provider_collateral_less_than_bound() {
+        let f = |rt: &mut MockRuntime, d: &mut DealProposal| {
+            // with these two equal provider collatreal min is 5/100 * deal size
+            let power = StoragePower::from_i128(1 << 50).unwrap();
```

Comment:
nit: this should be abstracted to something that can be shared by both this and the expect_query_network_info call.  This is what the harness is useful for in the go tests.  A refactor that size moving all market methods to an object is maybe too much here but you could put only expect_query_network_info on a queryable object and pass it into the closure.

Reply:
Will do this in the upcoming refactor.